src/dbnode/storage/bootstrap/bootstrapper/commitlog/source.go

```diff
@@ -336,44 +946,81 @@ func (s *commitLogSource) shouldIncludeInIndex(
 	return rangesToBootstrap.Overlaps(indexBlockRange)
 }
 
-func (s *commitLogSource) mergeShards(
+func (s *commitLogSource) mergeAllShardsCommitLogEncodersAndSnapshots(
+	ns namespace.Metadata,
+	shardsTimeRanges result.ShardTimeRanges,
+	snapshotFiles map[uint32]fs.FileSetFilesSlice,
+	mostRecentCompleteSnapshotByBlockShard map[xtime.UnixNano]map[uint32]fs.FileSetFile,
 	numShards int,
-	bopts result.Options,
 	blockSize time.Duration,
-	blopts block.Options,
-	encoderPool encoding.EncoderPool,
 	unmerged []shardData,
-) result.DataBootstrapResult {
+) (result.DataBootstrapResult, error) {
 	var (
-		shardErrs               = make([]int, numShards)
-		shardEmptyErrs          = make([]int, numShards)
-		bootstrapResult         = result.NewDataBootstrapResult()
-		blocksPool              = bopts.DatabaseBlockOptions().DatabaseBlockPool()
-		multiReaderIteratorPool = blopts.MultiReaderIteratorPool()
+		shardErrs       = make([]int, numShards)
+		shardEmptyErrs  = make([]int, numShards)
+		bootstrapResult = result.NewDataBootstrapResult()
 		// Controls how many shards can be merged in parallel
 		workerPool          = xsync.NewWorkerPool(s.opts.MergeShardsConcurrency())
 		bootstrapResultLock sync.Mutex
 		wg                  sync.WaitGroup
 	)
-
 	workerPool.Init()
 
 	for shard, unmergedShard := range unmerged {
+		snapshotData, err := s.bootstrapShardSnapshots(
+			ns.ID(),
+			uint32(shard),
+			false,
+			shardsTimeRanges[uint32(shard)],
+			blockSize,
+			snapshotFiles[uint32(shard)],
+			mostRecentCompleteSnapshotByBlockShard,
+		)
+		if err != nil {
+			bootstrapResultLock.Lock()
+			// Mark the shard time ranges as unfulfilled so a subsequent bootstrapper
+			// has the chance to fulfill it.
+			bootstrapResult.Add(
+				uint32(shard),
+				result.NewShardResult(0, s.opts.ResultOptions()),
+				shardsTimeRanges[uint32(shard)],
+			)
+			bootstrapResultLock.Unlock()
+			return nil, err
+		}
+
 		if unmergedShard.series == nil {
-			continue
+			if !snapshotData.IsEmpty() {
+				// No snapshot or commit log data, skip.
```

Comment:
Shouldn't this be `if snapshotData.IsEmpty() { ... }` then?  If this really a bug maybe we should add a test case too.